exploration/variable-mutability.md

```diff
@@ -0,0 +1,244 @@
+...
+modify = %6D.%6F.%64.%69.%66.%79
+```
+
+It is a syntax error to use `let` on a variable that has been previously
+assigned through any declaration (either `let` or `modify`)
+
+It is a variable resolution error to call `modify`
+on an external variable that does not exist.
+
+> _Example_
+>
+> ```
+> let @_local = {$external :transform}
+> modify @_local = {@_local :modification with=options}
+> modify $external = {$external :transform adding=annotation}
+> ```
+
+When more than one `modify` declaration applies to the same named variable,
+or when a `modify` declaration is applied to a local variable
+defined in a `let` declaration,
+the named variable behaves as if each declaration were called
+in the sequence in which they appear in the message.
+Implementations are not required (by this design, anyway)
+to resolve values in a greedy manner.
+They might not resolve a value unless it is actually used in a selector
+or in a placeholder.
+
+#### Sigil Choice for Local Variables
+
+The choice here of `@_` as the local variable sigil is probably not distinctive enough.
+It is probably okay to be a little inconvenient with local variable naming
+as these are less common than external variables.
+Alternatives to consider:
+
+- `@@foo`
+- `@foo@`
+- `@!foo`
+- `@ONLY_UPPER_ASCII_SNAKE`
+
+Note: if we have separate namespaces then local variables don't
+require Unicode names because their namespace is not subject
+to external data requirements.
+
+## Alternatives Considered
+
+_What other solutions are available?_
+_How do they compare against the requirements?_
+_What other properties they have?_
+
+### All Variables are Mutable; Shared Namespace
+
+**This is the current design.**
+A declaration can overwrite any passed in (external) value,
+either by adding annotation
+or by completely replacing the value.
+Further, one declaration can modify or completely overwrite a previous annotation.
+
+There are no warnings or errors produced when this occurs, even when it is unintentional.
+
+If variables are mutable and namespaces are shared, it's easy to write a message that never
+fails but does produce unintended or unexpected results (from the caller's point of view).
+
+```
+{"arg1": "10000"}
+...
+let $arg1 = {42}
+{This always says {$arg1} == 42}
+```
+
+### All Variables are Mutable; Non-shared Namespace
+
+If variables are mutable but namespaces are not shared, its easy for developers or translators to reference the wrong one:
+
+```
+{"arg1": "10000"}
+...
+let #arg1 = {42 :number maxFractionDigits=2}
+{This always says {$arg1} == 10000 because it should say {#arg1}}
+```
+
+### All Variables are Immutable; Shared Namespace
+
+If we make all variables immutable and external and local vars share a namespace,
+passing an argument that shares a name with a local declaration can cause a message to fail.
+
+```
+{ "arg1": "37"}
+...
+let $arg1 = {|42| :number maxFractionDigits=2}  // error
```

Comment:
Not throwing an error for this should be considered a requirement.

Reply:
In this option, this is an error, because one is attempting to overwrite `$arg1` with a new declaration. Presumably if we chose this option, it would be to **_ensure_** that this was an error.

I happen to agree that there is a use case that needs to be accounted for here (I have made the argument elsewhere many times about declarations not knowing what all the values are externally)